pkg/agent/controller/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,357 @@
+	basedir, err := afero.TempDir(defaultFS, "", "bundle_tmp_")
+	if err != nil {
+		return fmt.Errorf("error when creating temp dir: %w", err)
+	}
+	defer defaultFS.RemoveAll(basedir)
+
+	agentDumper := newAgentDumper(defaultFS, defaultExecutor, c.ovsCtlClient, c.aq, c.npq, supportBundle.SinceTime, c.v4Enabled, c.v6Enabled)
+	if err = agentDumper.DumpLog(basedir); err != nil {
+		return err
+	}
+	if err = agentDumper.DumpHostNetworkInfo(basedir); err != nil {
+		return err
+	}
+	if err = agentDumper.DumpFlows(basedir); err != nil {
+		return err
+	}
+	if err = agentDumper.DumpNetworkPolicyResources(basedir); err != nil {
+		return err
+	}
+	if err = agentDumper.DumpAgentInfo(basedir); err != nil {
+		return err
+	}
+	if err = agentDumper.DumpHeapPprof(basedir); err != nil {
+		return err
+	}
+	if err = agentDumper.DumpOVSPorts(basedir); err != nil {
+		return err
+	}
+
+	outputFile, err := afero.TempFile(defaultFS, "", "bundle_*.tar.gz")
+	if err != nil {
+		return fmt.Errorf("error when creating temp file: %w", err)
+	}
+	defer func() {
+		if err := outputFile.Close(); err != nil {
+			klog.Errorf("Error occurs when closing output tar file: %v", err)
+		}
+		if err := defaultFS.Remove(outputFile.Name()); err != nil {
+			klog.Errorf("Error occurs when removing output tar file %s: %v", outputFile.Name(), err)
+		}
+
+	}()
+	if _, err := compress.PackDir(defaultFS, basedir, outputFile); err != nil {
+		return fmt.Errorf("error when packaging support bundle: %w", err)
+	}
+
+	return c.uploadSupportBundle(supportBundle, outputFile)
+}
+
+func (c *SupportBundleController) uploadSupportBundle(supportBundle *cpv1b2.SupportBundleCollection, outputFile afero.File) error {
+	if _, err := outputFile.Seek(0, 0); err != nil {
+		return fmt.Errorf("support bundle upload to file server failed during setting offset: %v", err)
+
+	}
+	triesLeft := UPLOAD_TO_FILESERVER_TRIES
+	var uploadErr error
+	for triesLeft > 0 {
+		if uploadErr = c.uploadToFileServer(supportBundle.Name, &supportBundle.FileServer, &supportBundle.Authentication, outputFile); uploadErr != nil {
+			triesLeft -= 1
+			klog.Warningf("Failed to upload support bundle by SFTP: %v, triesLeft: %d", uploadErr, triesLeft)
+			time.Sleep(UPLOAD_TO_FILESERVER_RETRY_DELAY)
+		} else {
+			return nil
+		}
+	}
+	return fmt.Errorf("failed to upload support bundle by SFTP: %w", uploadErr)
+}
+
+func (c *SupportBundleController) uploadToFileServer(bundleName string, fileServer *cpv1b2.BundleFileServer, serverAuth *cpv1b2.BundleServerAuthConfiguration, tarGzFile io.Reader) error {
+	// fileServer.URL should be like: 10.92.23.154:22/path or sftp://10.92.23.154:22/path
+	var url string
+	if strings.Contains(fileServer.URL, "://") {
+		arr := strings.Split(fileServer.URL, "://")
+		if strings.ToLower(arr[0]) != "sftp" {
+			return fmt.Errorf("only sftp is supported for now, protocol %s is not supported", strings.ToLower(arr[0]))
+		}
+		url = fileServer.URL[7:]
```

Comment:
Better to calculate the index for url instead of hard-coded with 7 ?

Reply:
I thought about it. But the prefix length is fixed which is seven. If we use index, it would be like `strings.Index(fileServer.URL, "://") + 3`, which is not necessary in my mind.